builtin/logical/pki/path_manage_issuers.go

```diff
@@ -318,3 +320,167 @@ either take PEM-formatted certificates, and, if :type="bundle", unencrypted
 secret-keys.
 `
 )
+
+func pathRevokeIssuer(b *backend) *framework.Path {
+	fields := addIssuerRefField(map[string]*framework.FieldSchema{})
+
+	return &framework.Path{
+		Pattern: "issuer/" + framework.GenericNameRegex(issuerRefParam) + "/revoke",
+		Fields:  fields,
+
+		Operations: map[logical.Operation]framework.OperationHandler{
+			logical.UpdateOperation: &framework.PathOperation{
+				Callback: b.pathRevokeIssuer,
+				// Read more about why these flags are set in backend.go
+				ForwardPerformanceStandby:   true,
+				ForwardPerformanceSecondary: true,
+			},
+		},
+
+		HelpSynopsis:    pathRevokeIssuerHelpSyn,
+		HelpDescription: pathRevokeIssuerHelpDesc,
+	}
+}
+
+func (b *backend) pathRevokeIssuer(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	// Since we're planning on updating issuers here, grab the lock so we've
+	// got a consistent view.
+	b.issuersLock.Lock()
+	defer b.issuersLock.Unlock()
+
+	// Issuer revocation can't work on the legacy cert bundle.
+	if b.useLegacyBundleCaStorage() {
+		return logical.ErrorResponse("cannot revoke issuer until migration has completed"), nil
+	}
+
+	issuerName := getIssuerRef(data)
+	if len(issuerName) == 0 {
+		return logical.ErrorResponse("missing issuer reference"), nil
+	}
+
+	// Fetch the issuer.
+	sc := b.makeStorageContext(ctx, req.Storage)
+	ref, err := sc.resolveIssuerReference(issuerName)
+	if err != nil {
+		return nil, err
+	}
+	if ref == "" {
+		return logical.ErrorResponse("unable to resolve issuer id for reference: " + issuerName), nil
+	}
+
+	issuer, err := sc.fetchIssuerById(ref)
+	if err != nil {
+		return nil, err
+	}
+
+	// If its already been revoked, just return the read results sans warnings
+	// like we would otherwise.
+	if issuer.Revoked {
+		return respondReadIssuer(issuer)
+	}
+
+	// When revoking, we want to forbid new certificate issuance. We allow
+	// new revocations of leaves issued by this issuer to trigger a CRL
+	// rebuild still.
+	issuer.Revoked = true
+	if issuer.Usage.HasUsage(IssuanceUsage) {
+		issuer.Usage.ToggleUsage(IssuanceUsage)
+	}
+
+	currTime := time.Now()
+	issuer.RevocationTime = currTime.Unix()
+	issuer.RevocationTimeUTC = currTime.UTC()
+
+	err = sc.writeIssuer(issuer)
+	if err != nil {
+		return nil, err
+	}
+
+	// Rebuild the CRL to include the newly revoked issuer.
+	crlErr := b.crlBuilder.rebuild(ctx, b, req, false)
```

Comment:
@stevendpclark I think I either need to grab:

```
    b.revokeStorageLock.RLock()
    defer b.revokeStorageLock.RUnlock()
```

or I need to call `b.crlBuilder.requestRebuildIfActiveNode` -- preferably the former?

Reply:
Wait, I don't think we do. CRL Rebuilding is locked by an internal lock. This just locks the storage; issuer uses a separate set of storage and does not grab the issuers lock so I think we're good here.